Adding TEP for task results in final tasks

pritidesai opened the pull request:
Proposal for enabling final tasks to get value from non-final tasks.

[Design Doc](https://docs.google.com/document/d/10iEJqVstY6k3KNvAXgffIJLcHRbPQ-GIAfQk5Dlrf3c/edit#heading=h.rghl981jtej4)

Issue [#2557](https://github.com/tektoncd/pipeline/issues/2557)

bobcatfish commented on teps/draft-20200716-task-results-in-final-tasks.md:
```diff
@@ -0,0 +1,279 @@
+
+A `Pipeline` author can send `Pull Request` ID to finally task after building the source code and
+running unit/integration tests. The finally `Task` can send Slack notification with that ID and
+the result of the build.
```
I wonder if it would make sense to replace this with a user story about clean up; in the motivation section we only discuss cleanup, and this particular user story has some other questions, since a finally task that is notifying might want to run even when the task it depends on fails (to update about the failure)

pritidesai replied:
Done 👍

bobcatfish commented on teps/draft-20200716-task-results-in-final-tasks.md:
```diff
@@ -0,0 +1,279 @@
+
+* No API changes.
+
+* Finally contract does not break and all final tasks are attempted.
```
👍  this was a great call, thanks for thinking of this approach @pritidesai !

bobcatfish commented on teps/draft-20200716-task-results-in-final-tasks.md:
```diff
@@ -0,0 +1,279 @@
+      default: "best-default-ever"
+      schema:
+        type: string
+        minLen: 1
```
i'm a bit confused by this example: if the default value of the param will be used, I don't think the schema validation will come into play at all?

pritidesai replied:
I can drop the `schema` from here for now, but the param validation must be done on the defaults as well once we introduce `schema`. And such param validation happens before scheduling any tasks in the pipeline for execution. Its considered more of a syntax validation i.e. part of compilation (if it make sense 😉 )

pritidesai commented:
@bobcatfish I have addressed your comments and also added a very high level [workflow](https://github.com/tektoncd/community/pull/153/files#diff-582a69116e2e369bd947825b6ab89ff6R160) to help understand it better.

What are the next steps in terms of changing the filename or status? 🤔

pritidesai commented:
Also, two things I came across while working on this PR regarding the `TEP` template in general:

1. Links to GitHub issue and/or design doc in Google drive - should we create a separate section on `References` or add a placeholder to insert the GitHub issue? or something similar?

2. Inserting images, might be a good idea to designate a directory called `/teps/images` and  ask `TEP` authors to add files in that directory by following a specific naming convention. 

Thoughts?

vdemeester commented:
>     1. Links to GitHub issue and/or design doc in Google drive - should we create a separate section called `References` or add a placeholder to insert the GitHub issue? or something similar?

Good point, I think we can definitely have a `References` header to help referering to anything we may want to refer (other TEPs, design docs, issues, …)

>     2. Inserting images, might be a good idea to designate a directory called `/teps/images/` and  ask `TEP` authors to add files in that directory by following a specific naming convention.

Either that or `/teps/TEP-0004/` for TEP-0004 resources (images, …) ? Not sure which one I prefer :stuck_out_tongue_closed_eyes: 

> @vdemeester @bobcatfish Thoughts?

vdemeester commented on teps/draft-20200716-task-results-in-final-tasks.md:
```diff
@@ -0,0 +1,304 @@
+If the `Task` producing the value fails, the finally `Task` that requires that value will be marked as failed.
+That final `Task` will be `attempted` i.e. evaluated by the `Pipeline Controller` for execution,
+but the validation will fail before the `TaskRun` is created since the required `Result` is not available.
+However if the `Task` being attempted has a default value for that parameter, it can still run,
+using the default when result is not available.
```
I would highly a bit more the `default` part. In a few sentences what we are proposing is : You can refer to results from tasks, and provide a default value if those results did not get populated. Any non-populated results that do not have a default will make the finally task fail.

vdemeester replied:
(oh it's on the "default" params, I see)

vdemeester commented on teps/draft-20200716-task-results-in-final-tasks.md:
````diff
@@ -0,0 +1,304 @@
+```
+
+**Note:** Today final tasks are executed all in parallel and can not depend on any other `Task` in the `Pipeline`.
+With adding support for `Task` `Results`, we are introducing dependencies to finally `Tasks` but that
````
I am not sure I understand this (nor if it should be there). What we are saying is "now finally tasks depends on tasks" ? that was implicitly the case already (from the user perpective) as the finally task will wait for all the "normal" task to complete (success or failure).

pritidesai replied:
> What we are saying is "now finally tasks depends on tasks" ? that was implicitly the case already (from the user perpective) as the finally task will wait for all the "normal" task to complete (success or failure).

Finally tasks did not have any implicit/explicit dependency on any other task. The scheduling algorithm sends final tasks to execute after all normal/DAG tasks are completed. Now, with task results, implicit dependency is being introduced. There is no impact on scheduling yet, final tasks still scheduled after all DAG. But worth calling out now and will easier to understand what happens when we add support for task results from other final tasks, `runAfter` in final tasks, `from` in final tasks, etc

bobcatfish commented:
> Good point, I think we can definitely have a References header to help referering to anything we may want to refer (other TEPs, design docs, issues, …)

great idea!!

> 2. Inserting images, might be a good idea to designate a directory called `/teps/images/` and  ask `TEP` authors to add files in that directory by following a specific naming convention.
> Either that or /teps/TEP-0004/ for TEP-0004 resources (images, …) ? Not sure which one I prefer 😝

i have no strong opinion! slight inclination toward `/teps/images/` to keep it simple

bobcatfish commented:
> What are the next steps in terms of changing the filename or status? 🤔

I'm thinking we can at least merge as "proposed" asap (correct me if I'm wrong @vdemeester ) with the final filename, and then open another PR to change that to "implementable" and continue discussing

bobcatfish commented on teps/draft-20200716-task-results-in-final-tasks.md:
````diff
@@ -0,0 +1,304 @@
+
+**A.**
+
+```
````
can you explain the context of this workflow a bit more? e.g. " Param Validation " i'm not sure if you mean, params to the finally tasks? params for all tasks?

(also im confused how we can validate the params before we apply results from dependent tasks?)

pritidesai replied:
In this context, "Param validation" does syntax check on params to all final tasks:

(1) check if parameter name is unique
(2) type of the default value matches with its type
(3) param with task results has valid result reference i.e. $(tasks.results.taskName.param)

"Validate and Apply results from Dependent Tasks" actually resolves the task result from dependent tasks and applies those values to params in the final tasks.

In general, the same workflow is executed for non-final tasks too.

bobcatfish replied:
k i think this is a bit more clear to me

bobcatfish commented on teps/draft-20200716-task-results-in-final-tasks.md:
```diff
@@ -0,0 +1,304 @@
+
+* This proposal is extensible with a param validation proposal using Schema -
+[#1393](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-558833526) by reading default from
+`Task` specification.
```
would it make sense to remove this reference to schema as well? maybe this is a discussion for somewhere else, but this is how i would see this feature interacting with schema:

- Param defaults should always pass their own schema; if a task declares a default that does not pass it's schema, the Task creation should fail
- If we one day add Result defaults, when a Task fails, the default of the result could be used as the Result of that Task. If a Task wants to use that result as the value for one if its own params, then the Task validation will evaluate that value; if the default is invalid, the validation will fail, if it is valid, the Task can continue

pritidesai replied:
sure @bobcatfish, shall I drop the whole bullet point along with the example?

pritidesai replied:
never mind, i just dropped the reference and have the example as is.

bobcatfish commented on teps/draft-20200716-task-results-in-final-tasks.md:
```diff
@@ -0,0 +1,304 @@
+
+> `PipelineRun` reads parameter defaults from `Pipeline` specification but in case of
+parameters associated with `Task` `Results`, the default values will be used from the `Task` since
+`Pipeline` has no direct mapping between `Pipeline` parameter and `Task` parameter.
```
is the example below meant to demonstrate that even if `boskos-acquire` fails, `boskos-release` will still run because it has a default for the param? if so i think we might want to say that explicitly

pritidesai replied:
yup, added it explicitly 😄

vdemeester commented:
So we should "assign" a number to the TEP before merging as proposed (so that we rename it `0005-…`), that would be the only requirement :)
Feels like it should be `0004` as I see this one getting merged before https://github.com/tektoncd/community/pull/147 for example :upside_down_face:

pritidesai commented:
@bobcatfish @vdemeester I have renamed the file, made two changes:

1. Dropped the reference to GitHub issue with param validation schema
2. Added explicitly "task is executed with default value if dependency fails and no task result produced" under "Advantages" section.

Ready to review again 🙏

bobcatfish commented:
Let's go for it! Then we can open a PR to get it to implementable :D

/approve
